File: design-patterns/ts/template-method/transport/plane.ts
// plane.ts

import Transport from "./transport";


class Plane extends Transport {
    createBody(): void {
        console.log("Creating Plane Body");
    }

    addEngine(): void {
        console.log("Adding Engine to Plane");
    }

    addWheel(): void {
        console.log("Adding 3 Wheels to Plane");
    }

    addWing(): void {
        console.log("Adding Wings Plane");
    }
}

export default Plane;

File: design-patterns/go/template_method/transport/plane.go
// plane.go

package main

import "fmt"

type Plane struct {
	*Transport
}

func NewPlane() (plane *Plane) {
	plane = &Plane{}
	plane.Transport = NewTransport(plane)
	return
}

func (plane *Plane) AddEngine() {
	fmt.Println("Adding Engine to Plane")
}

func (plane *Plane) AddWheel() {
	fmt.Println("Adding 3 Wheels to Plane")
}

func (plane *Plane) AddWing() {
	fmt.Println("Adding Wings Plane")
}

func (plane *Plane) CreateBody() {
	fmt.Println("Creating Plane Body")
}
